edge-hub/src/Microsoft.Azure.Devices.Routing.Core/endpoints/StoringAsyncEndpointExecutor.cs

```diff
@@ -159,22 +167,22 @@ public async Task UpdatePriorities(IList<uint> priorities, Option<Endpoint> newE
                     EndpointExecutorFsm fsm = new EndpointExecutorFsm(this.Endpoint, checkpointer, this.config);
 
                     // Add it to our dictionary
-                    snapshot.Add(priority, fsm);
+                    updatedSnapshot.Add(priority, fsm);
                 }
                 else
                 {
                     // Update the existing FSM with the new endpoint
-                    EndpointExecutorFsm fsm = snapshot[priority];
+                    EndpointExecutorFsm fsm = updatedSnapshot[priority];
                     await newEndpoint.ForEachAsync(e => fsm.RunAsync(Commands.UpdateEndpoint(e)));
                 }
             }
 
-            this.prioritiesToFsms.Value = snapshot;
-            Events.UpdatePrioritiesSuccess(this, snapshot.Keys.ToList());
+            this.prioritiesToFsms.CompareAndSet(snapshot, updatedSnapshot.ToImmutableDictionary());

```

Comment:
What happens if this compare fails?

Reply:
There are 2 places calling prioritiesToFsms.CompareAndSet in this class, which are UpdatePriorites and Dispose methods.  Dispose method should be called by GC once CloseAsync method is called and free up all the references and all other methods are completed.  Therefore Dispose should not be called at the same time of UpdatePriorities.  So what happens if 2 threads calling UpdatePriorites method?  Currently there are 2 places call to UpdatePriorities method, 1) from StoringAsyncEndpointExecutorFactory.CreateAsync method, this is create scenario.  Therefore it should not have 2 threads to try calling UpdatePriorities. 2) from its SetEndpoint method, this means 2 config change requests sending to EdgeHub almost at the same time and one is updated successfully but not the other.  This line will throw exception and will catch in SetEndpoint method try-catch block and report SetEndpointFailure, then rethrow and caught again in ConfigUpdater, which logs an error.

Reply:
Yes. I will log an event if CompareAndSet returns false.  I traced up the usages and all methods changing endpoints in dispatcher used async lock to avoid 2 threads to calling any executor at the same time.

Reply:
In that case, why do you need CompareAndSet? Just use the Set.

Reply:
It is more preventive.  Even caller get changed and causes race condition, at least we will report it in log.